native/src/main/java/io/ballerina/lib/avro/Avro.java

```diff
@@ -52,46 +59,32 @@ public static void generateSchema(BObject schemaObject, BString schema) {
 
     public static Object toAvro(BObject schemaObject, Object data) {
         Schema schema = (Schema) schemaObject.getNativeData(AVRO_SCHEMA);
-        ObjectMapper objectMapper = new ObjectMapper();
-        try {
-            Object jsonObject = generateJsonObject(data, schema, objectMapper);
-            if (jsonObject == null) {
-                return ValueCreator.createArrayValue(new byte[]{0, 0});
-            } else if (Objects.equals(schema.getType(), Schema.Type.FIXED)) {
-                return ValueCreator.createArrayValue(((BArray) jsonObject).getByteArray());
-            }
-            byte[] avroBytes = (new AvroMapper()).writer(new AvroSchema(schema)).writeValueAsBytes(jsonObject);
-            return ValueCreator.createArrayValue(avroBytes);
-        } catch (JsonProcessingException e) {
-            return Utils.createError(JSON_PROCESSING_ERROR, e);
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+            SerializeVisitor serializeVisitor = new SerializeVisitor();
+            Serializer serializer = MessageFactory.createMessage(schema);
+            Object avroData = Objects.requireNonNull(serializer).generateMessage(serializeVisitor, data);
+            DatumWriter<Object> writer = new GenericDatumWriter<>(schema);
+            BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(outputStream, null);
+            writer.write(avroData, encoder);
+            encoder.flush();
+            byte[] bytes = outputStream.toByteArray();
+            return ValueCreator.createArrayValue(bytes);
```

Comment:
No need of creating a separate variable here.

```suggestion
            return ValueCreator.createArrayValue(outputStream.toByteArray());
```